spring-test/src/test/kotlin/org/springframework/test/web/servlet/MockMvcDslTests.kt

```diff
@@ -0,0 +1,150 @@
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.ResponseBody
+
+import org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup
+
+/**
+ * MockMvcDsl tests that verify builder, actions, and expect blocks
+ *
+ * @author Clint Checketts
+ * @author Petr Balat
+ */
+class MockMvcDslTests {
+	private lateinit var mockMvc: MockMvc
+
+	@Before
+	fun setup() {
+		mockMvc = standaloneSetup(PersonController()).build()
+	}
+
+	@Test
+	fun json() {
+		mockMvc.perform(HttpMethod.GET,"/person/{name}", "Lee") {
+			builder { accept(MediaType.APPLICATION_JSON) }
+			print()
+			expect {
+				status { isOk }
+				content { contentType("application/json;charset=UTF-8") }
+				jsonPath("$.name") { value("Lee") }
+				json("""{"someBoolean": false}""", strict = false)
+			}
+		}
+	}
+
+	@Test
+	fun `negative assertion tests to verify the matchers throw errors when expected`() {
+		val name = "Petr"
+		mockMvc.perform(HttpMethod.GET,"/person/$name") {
+			builder { accept(MediaType.APPLICATION_JSON) }
+			print()
+			expect {
+				assertThrows<AssertionError> { content { contentType(MediaType.APPLICATION_ATOM_XML) } }
+				assertThrows<AssertionError> { contentString("Wrong") }
+				assertThrows<AssertionError> { jsonPath("name", CoreMatchers.`is`("Wrong")) }
+				assertThrows<AssertionError> { json("""{"name":"wrong"}""") }
+				assertThrows<AssertionError> { jsonPath("name") { value("wrong") } }
+				assertThrows<AssertionError> { cookie { value("name", "wrong") } }
+				assertThrows<AssertionError> { flash { attribute<String>("name", "wrong") } }
+				assertThrows<AssertionError> { header { stringValues("name", "wrong") } }
+				assertThrows<AssertionError> { model { attributeExists("name", "wrong") } }
+				assertThrows<AssertionError> { model<String>("name") { Assert.assertThat(this, CoreMatchers.`is`("wrong")) } }
+				assertThrows<AssertionError> { redirectedUrl("wrong/Url") }
+				assertThrows<AssertionError> { redirectedUrlPattern("wrong/Url") }
+				assertThrows<AssertionError> { redirectedUrlPattern("wrong/Url") }
+				assertThrows<AssertionError> { status { isAccepted } }
+				assertThrows<AssertionError> { viewName("wrongName") }
+				assertThrows<AssertionError> { HttpStatus.ACCEPTED.isStatus() }
+				assertThrows<AssertionError> { "$.name" jsonPathIs "wrong" }
+				assertThrows<AssertionError> { "$.name" jsonPathMatcher CoreMatchers.`is`("wrong") }
+				assertThrows<AssertionError> { "name" jsonPath { value("wrong") } }
+				assertThrows<AssertionError> { jsonPath("name") { value("wrong") } }
+				assertThrows<AssertionError> { +HandlerMethod("helloJsonWrong") }
+			}
+		}
+	}
+
+	@Test
+	fun `negative assertion tests for xpath`() {
+		mockMvc.perform(HttpMethod.GET,"/person/Clint") {
+			builder { accept(MediaType.APPLICATION_XML) }
+
+			andDo(MockMvcResultHandlers.print())
+			andExpect(status().isOk)
```

Comment:
I left `andDo` and `andExpect` as methods that allow the user to transition to the new syntax. They can use the syntax they are familiar with as well as include their existing matchers.

By changing to `print()` for printing the request and response then it forces the user to qualify `MockMvcResultHandlers.print()`, but I think that is fine.

I haven't switched `andDo` to `handle` yet. I think we need some better examples before changing that. `handle(document())` for RestDocs for example feels weird.